File: .reasonix/skills/abacatepay/tests/go/webhook_test.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/http"
	"net/http/httptest"
	"os"
	"testing"
)

func TestWebhookHandler(t *testing.T) {
	// Set up test data
	secret := "test_secret"
	os.Setenv("ABACATEPAY_WEBHOOK_SECRET", secret)
	defer os.Unsetenv("ABACATEPAY_WEBHOOK_SECRET")

	body := `{"event": "test"}`
	bodyBytes := []byte(body)

	// Calculate expected signature
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(bodyBytes)
	expectedSignature := hex.EncodeToString(mac.Sum(nil))

	// Create request
	req := httptest.NewRequest("POST", "/webhook?webhookSecret="+secret, bytes.NewReader(bodyBytes))
	req.Header.Set("X-Webhook-Signature", expectedSignature)
	req.Header.Set("Content-Type", "application/json")

	// Create response recorder
	w := httptest.NewRecorder()

	// Call handler
	webhookHandler(w, req)

	// Check response
	if w.Code != http.StatusOK {
		t.Errorf("Expected status 200, got %d", w.Code)
	}

	if w.Body.String() != "OK" {
		t.Errorf("Expected body 'OK', got %s", w.Body.String())
	}
}

func TestWebhookHandlerInvalidSecret(t *testing.T) {
	os.Setenv("ABACATEPAY_WEBHOOK_SECRET", "secret")
	defer os.Unsetenv("ABACATEPAY_WEBHOOK_SECRET")

	req := httptest.NewRequest("POST", "/webhook?webhookSecret=wrong", nil)
	w := httptest.NewRecorder()

	webhookHandler(w, req)

	if w.Code != http.StatusUnauthorized {
		t.Errorf("Expected status 401, got %d", w.Code)
	}
}

func TestWebhookHandlerInvalidSignature(t *testing.T) {
	secret := "secret"
	os.Setenv("ABACATEPAY_WEBHOOK_SECRET", secret)
	defer os.Unsetenv("ABACATEPAY_WEBHOOK_SECRET")

	body := `{"event": "test"}`
	req := httptest.NewRequest("POST", "/webhook?webhookSecret="+secret, bytes.NewReader([]byte(body)))
	req.Header.Set("X-Webhook-Signature", "invalid")

	w := httptest.NewRecorder()

	webhookHandler(w, req)

	if w.Code != http.StatusUnauthorized {
		t.Errorf("Expected status 401, got %d", w.Code)
	}
}

File: .reasonix/skills/abacatepay/tests/ts/webhook.test.ts
import express from 'express';
import request from 'supertest';
import crypto from 'crypto';

// Assuming the webhook handler is exported or we can test the route
// For this example, we'll simulate the app setup

describe('Webhook Handler', () => {
  let app: express.Application;
  const secret = 'test_secret';

  beforeEach(() => {
    app = express();
    app.use(express.json({
      verify: (req: any, res, buf) => {
        req.rawBody = buf;
      }
    }));

    app.post('/webhooks/abacatepay', (req, res) => {
      const signature = req.headers['x-webhook-signature'];
      const secretFromQuery = req.query.webhookSecret;

      if (secretFromQuery !== secret) {
        return res.status(401).send('Invalid secret query');
      }

      const hmac = crypto.createHmac('sha256', secret);
      hmac.update(req.rawBody);
      const expectedSignature = hmac.digest('base64');

      if (signature !== expectedSignature) {
        return res.status(401).send('Invalid signature');
      }

      const { event: eventType, data } = req.body;
      console.log('Received event:', eventType);

      if (eventType === 'billing.paid') {
        console.log('Payment confirmed for:', data.id);
      }

      res.status(200).send('OK');
    });
  });

  test('should accept valid webhook', async () => {
    const body = { event: 'billing.paid', data: { id: '123' } };
    const bodyString = JSON.stringify(body);
    const hmac = crypto.createHmac('sha256', secret);
    hmac.update(Buffer.from(bodyString));
    const signature = hmac.digest('base64');

    const response = await request(app)
      .post('/webhooks/abacatepay?webhookSecret=' + secret)
      .set('x-webhook-signature', signature)
      .set('Content-Type', 'application/json')
      .send(body);

    expect(response.status).toBe(200);
    expect(response.text).toBe('OK');
  });

  test('should reject invalid secret', async () => {
    const response = await request(app)
      .post('/webhooks/abacatepay?webhookSecret=wrong')
      .send({});

    expect(response.status).toBe(401);
    expect(response.text).toBe('Invalid secret query');
  });

  test('should reject invalid signature', async () => {
    const response = await request(app)
      .post('/webhooks/abacatepay?webhookSecret=' + secret)
      .set('x-webhook-signature', 'invalid')
      .send({});

    expect(response.status).toBe(401);
    expect(response.text).toBe('Invalid signature');
  });
});